### ISSUE TITLE
Could NOT start training on RTX A6000 but runs well on TITAN RTX and 2080ti 

### ISSUE THREAD

**Issue by RossStream** [pytorch/pytorch#76101]
### 🐛 Describe the bug

Hello, 

We meet a problem when we try to train our model on NVIDIA RTX A6000 GPU server.
The same code runs smoothly on GeForce RTX 2080 Ti and TITAN RTX servers.

We use totally the same conda virtual env for all servers, but after we run our code by "python train.py", normally on 2080 Ti and TITAN RTX, all the parameters of our model would be load to the GPUs (take like 32G GPU memory) and the training would start, we would know the process of training and estimated finishing time:
`loss: 3.49e+04 :   0%|          | 11/180060 [00:32<81:27:03,  1.16it/s]`
`loss: 3.49e+04 :   0%|          | 12/180060 [00:33<73:16:11,  1.29it/s]`

But on RTX A6000, the parameters could also be loaded to the GPUs and take 32G GPU memory, but the training could NOT start and we could only get this (it keeps here in almost 12 hours):
`0%|          | 0/180060 [00:00<?, ?it/s]`

I check the environment by `collect_env.py` and following is a comparison for our TITAN RTX server, 2080 Ti Server and RTX A6000 server.

### TITAN RTX Server
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-105-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: NVIDIA TITAN RTX
GPU 1: NVIDIA TITAN RTX

Nvidia driver version: 470.74
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.5           py39he7a7128_1  
[conda] numpy-base                1.21.5           py39hf524024_1  
[conda] pytorch                   1.11.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.11.0               py39_cu113    pytorch
[conda] torchvision               0.12.0               py39_cu113    pytorch

### 2080 Ti Server 
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-163-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.4.100
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 460.106.00
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.5           py39he7a7128_1  
[conda] numpy-base                1.21.5           py39hf524024_1  
[conda] pytorch                   1.11.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.11.0               py39_cu113    pytorch
[conda] torchvision               0.12.0               py39_cu113    pytorch

### RTX A6000 server
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-91-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000

Nvidia driver version: 470.86
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.5           py39he7a7128_1  
[conda] numpy-base                1.21.5           py39hf524024_1  
[conda] pytorch                   1.11.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.11.0               py39_cu113    pytorch
[conda] torchvision               0.12.0               py39_cu113    pytorch

cc @ngimel

**Comment by ptrblck** [comment]
@RossStream could you post a minimal, executable code snippet to reproduce the issue, please?

**Comment by RossStream** [comment]
@ptrblck 
### Demo Code
Here is the code (size: ~200 KB): [Google drive](https://drive.google.com/drive/folders/1sTNkbwh3g6vQn41VSYs8g4YdJeqgAdo2?usp=sharing)

Includes:
```
./train
./dev
train.py
```

### Environment
```
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch
pip install timm
conda install tensorboard
conda install tqdm
```

### Start training
`python train.py`

**Comment by RossStream** [comment]
@ptrblck @ngimel 

Any update on this issue?
I searched for many solutions but still could not solve this and I am stuck here for over a week ...

Thx.

**Comment by RossStream** [comment]
Hello all,

@ptrblck @ngimel 

Any update? :)

Thx!

**Comment by RossStream** [comment]
No update since Apr 20, so I close the issue.

**Comment by mjun0812** [comment]
I have the same issue, but I've solved it.

You can set this environment value.

```bash
export NCCL_P2P_DISABLE=1
```

This value's info is [here](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-p2p-disable) and also discussed [here](https://discuss.pytorch.org/t/single-machine-ddp-issue-on-a6000-gpu/134869/6).

Before setting the environment value, My DDP training script worked 2 x Geforce RTX 3090 but didn't work 2 x RTX A6000.

My workstation Info is below.

```bash
Collecting environment information...
PyTorch version: 1.11.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.10 (main, Jan 27 2022, 02:22:40)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-113-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA RTX A6000
GPU 3: NVIDIA RTX A6000

Nvidia driver version: 510.73.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.11.0+cu113
[pip3] torchaudio==0.11.0+cu113
[pip3] torchinfo==1.7.0
[pip3] torchvision==0.12.0+cu113
[conda] Could not collect
```

**Comment by dovletov** [comment]
I recently encountered a similar problem with the following machine:
AMD Ryzen Threadripper PRO 3975WX 32-Cores
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000
I was not able to train my networks in the multi-GPU setting since the training routine was freezing at the first training iteration.
However, the same code was working without any problems when choosing a single-GPU setting.
Data parallelization was implemented using `nn.DataParallel` 

I also tried different CUDA versions (10.1, 11.3) and different Nvidia driver versions (495, 510, 515), however, was not able to fix this issue.

As suggested by @mjun0812 in this thread, I tried to fix this issue by setting the environmental variable `export NCCL_P2P_DISABLE=1`. This worked in the sense that the training started without any freezing when the multi-GPU setting was chosen. However, the performance of my image-to-image translation networks degraded drastically in comparison to the results that were obtained using another server with 2 x NVIDIA GeForce GTX 1080 Ti. 

After starting the same experiment in these two different machines, I noticed that my validation losses are not in the same range. Moreover, one of the networks was converging much faster than another. After a visual inspection of the validation results (from RTX A6000 GPUs), I noticed that only part of them is correct. In each stack of results within a single batch iteration, the first 8 images were making some sense, while the following 8 were completely wrong. Both experiments were done using the same docker image with batch size set to 16.

In the end, I solved my problem by disabling [IOMMU] as suggested in https://github.com/pytorch/pytorch/issues/52142
